consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+
+        match &block.prev_block_id() {
+            Some(block_id) => {
+                let previous_block = self
+                    .db_tx
+                    .get_block_index(&Id::<Block>::new(&block_id.get()))?
+                    .ok_or(BlockError::Orphan)?;
+                // Time
+                let block_time = block.block_time();
+                if previous_block.get_block_time() > block_time {
+                    return Err(BlockError::Unknown);
+                }
+                if i64::from(block_time) > time::get() {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            None => {
+                // This is only for genesis, AND should never come from a peer
+                if block_source != BlockSource::Local {
+                    return Err(BlockError::InvalidBlockSource);
+                };
+            }
+        }
+
+        self.check_transactions(block)?;
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
+        let _consensus_data = block.consensus_data();
+        // TODO: PoW is not in master at the moment =(
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_transactions(&self, block: &Block) -> Result<(), BlockError> {
+        // TODO: Must check for duplicate inputs (see CVE-2018-17144)
+        //      We should discuss - can we add Hash trait to Transaction?
+        //      We will have plenty more checks with inputs\outputs and HashSet\BTreeMap might be more efficient
+        //
+        // let mut keyed = HashSet::new();
+        // for tx in block.get_transactions() {
+        //     for input in tx.get_inputs() {
+        //         if keyed.contains(input.get_outpoint()) {
+        //             return Err(BlockError::Unknown);
+        //         }
+        //         keyed.insert(input.get_outpoint());
+        //     }
+        // }
+
+        //TODO: Size limits
+        if Encode::encoded_size(block) > MAX_BLOCK_WEIGHT {
+            return Err(BlockError::Unknown);
+        }
+        //TODO: Check signatures will be added when will ready BLS
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_block(&self, block: &Block, block_source: BlockSource) -> Result<(), BlockError> {
+        self.check_consensus(block)?;
+        self.check_block_detail(block, block_source)?;
+        Ok(())
+    }
+
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> bool {
+        block_index.get_next_block_id().is_some()
+            || self
+                .db_tx
+                .get_best_block_id()
+                .ok()
+                .flatten()
+                .map_or(false, |ref block_id| block_index.get_block_id() == block_id)
+    }
+
+    /// Mark new block as an orphan
+    fn new_orphan_block(&mut self, block: Block) -> Result<(), BlockError> {
+        if block.prev_block_id().is_none() && block.is_genesis(self.chain_config) {
```

Comment:
This is wrong. It should be ||. Besides, this should be impossible as we should check before whether this is genesis. Attempting to add genesis or an invalid block to orphans is a logic error in our implementation. Just assert this stuff.

Reply:
Ohh, thanks!

> Just assert this stuff

Did you mean something like it: 
```rust        
assert!(block.prev_block_id().is_some() && !block.is_genesis(self.chain_config));
```

Reply:
Due to I added check about `prev_block_id` to `check_block_detail`, we can leave just: 
```rust
        assert!(!block.is_genesis(self.chain_config));
```